### PR TITLE
[SPARK-41434][CONNECT][PYTHON] Initial `LambdaFunction` implementation

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
There are 11 lambda functions, this PR adds the basic support for `LambdaFunction` and add the  `exists` function.

### Why are the changes needed?
for API coverage

### Does this PR introduce _any_ user-facing change?
yes, new API

### How was this patch tested?
added UT

### CODE REVIEW

**Comment by zhengruifeng** [comment]
reviewers can refer to the implementation in PySpark

https://github.com/apache/spark/blob/master/python/pyspark/sql/functions.py#L8094-L8197

File: python/pyspark/sql/connect/functions.py [hunk 1]
```diff
@@ -79,6 +85,84 @@ def _invoke_binary_math_function(name: str, col1: Any, col2: Any) -> Column:
...
+            - (Column, Column) -> Column: ...
+            - (Column, Column, Column) -> Column: ...
+    """
+    parameters = _get_lambda_parameters(f)
+
+    arg_names = ["x", "y", "z"]
+
+    arg_cols: List[Column] = []
+    for arg in arg_names[: len(parameters)]:
+        # TODO: How to make sure lambda variable names are unique? RPC for increasing ID?
```
Comment (zhengruifeng) [inline python/pyspark/sql/connect/functions.py:132]: PySpark invokes `UnresolvedNamedLambdaVariable.freshVarName` in JVM to get a unique variable name

 https://github.com/apache/spark/blob/master/python/pyspark/sql/functions.py#L8157

```
object UnresolvedNamedLambdaVariable {

  // Counter to ensure lambda variable names are unique
  private val nextVarNameId = new AtomicInteger(0)

  def freshVarName(name: String): String = {
    s"${name}_${nextVarNameId.getAndIncrement()}"
  }
}
```

**Comment by zhengruifeng** [comment]
The failed `./dev/test-dependencies.sh` is irrelevant

**Comment by zhengruifeng** [comment]
cc @HyukjinKwon @cloud-fan @amaliujia @grundprinzip @hvanhovell

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto [hunk 2]
```diff
@@ -167,4 +169,26 @@ message Expression {
...
+    // (Required) Function arguments. Must not be empty.
+    //
+    // All arguments must be UnresolvedNamedLambdaVariables.
+    repeated Expression arguments = 2;
+
+    // (Required) A lambda function can be hidden when a user wants to
+    // process an completely independent expression in a 'HigherOrderFunction',
+    // the lambda function and its variables are then only used for internal
+    // bookkeeping within the higher order function.
+    bool hidden = 3;
```
Comment (hvanhovell) [inline connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:186]: Can we see if we can do without this one? This is very much an implementation detail.

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto [hunk 3]
```diff
@@ -167,4 +169,26 @@ message Expression {
...
+  }
+
+  message UnresolvedNamedLambdaVariable {
```
Comment (hvanhovell) [inline connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:189]: Can we just make this an unresolved attribute? And do the heavy lifting in the connect planner?

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto [hunk 4]
```diff
@@ -167,4 +169,26 @@ message Expression {
     // (Optional) Alias metadata expressed as a JSON map.
     optional string metadata = 3;
   }
+
+  message LambdaFunction {
```
Comment (hvanhovell) [inline connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:173]: I am on the fence about adding this message. We could go the CaseWhen route here as well. The only argument against that are the arguments them self :).

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto [hunk 5]
```diff
@@ -167,4 +169,26 @@ message Expression {
...
+    // (Required) The lambda function.
+    Expression function = 1;
+
+    // (Required) Function arguments. Must not be empty.
+    //
+    // All arguments must be UnresolvedNamedLambdaVariables.
+    repeated Expression arguments = 2;
```
Comment (amaliujia) [inline connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:180]: If all arguments must be UnresolvedNamedLambdaVariables. why not define this as 
`repeated UnresolvedNamedLambdaVariables arguments = 2`?

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto (hunk 3 above)
Comment (amaliujia) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:189]: Unless there is special resolution in Catalyst on this type otherwise +1 to re-use `unresolved attribute`

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto (hunk 5 above)
Comment (zhengruifeng) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:180]: I was hitting something weird if using `UnresolvedNamedLambdaVariables ` here

**Review by grundprinzip — COMMENTED** [review]
First pass.

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto (hunk 3 above)
Comment (hvanhovell) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:189]: it are literally the argument names in a lambda, for example `x => x + 1`, the `UnresolvedNamedLambdaVariable` is `x`.

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto (hunk 2 above)
Comment (zhengruifeng) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:186]: will remove it.

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto (hunk 4 above)
Comment (zhengruifeng) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:173]: > How would the lambda function look like for using the `CaseWhen` route?

this one https://github.com/apache/spark/pull/38956, we build the expressions in Connect Planner other than let `FunctionRegistry` do this

Comment (zhengruifeng) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto]: actually, we can remove `ExpressionString`, `UnresolvedStar`, `UnresolvedRegex`, `Cast` in the same way.

But now I'm not sure whether it is the correct way.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 6]
```diff
@@ -565,6 +565,47 @@ class SparkConnectPlanner(session: SparkSession) {
         val children = fun.getArgumentsList.asScala.toSeq.map(transformExpression)
         Some(In(children.head, children.tail))
 
+      case "___lambda_function___" =>
+        // UnresolvedFunction[___lambda_function___, ["x, y -> x < y", "x", "y"]]
```
Comment (grundprinzip) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:569]: This branch deserves it's own function please.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 7]
```diff
@@ -565,6 +565,47 @@ class SparkConnectPlanner(session: SparkSession) {
...
+
+        val function = children.head
+
+        val variableNames = children.tail.map {
+          case variable: UnresolvedAttribute if variable.nameParts.length == 1 =>
+            variable.nameParts.head
+          case other =>
+            throw InvalidPlanInput(
+              "LambdaFunction requires all arguments to be UnresolvedAttribute with " +
+                s"single name part, but got $other")
```
Comment (grundprinzip) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:586]: There are two interesting issues here:

 1. When someone submits to the API an expression that does not transform into `UnresolvedExpression` this would throw a weird error message about the name parts but actually the type does not match.
 2. Why the restriction to single part names? Is this a Spark limitation?

File: python/pyspark/sql/connect/column.py [hunk 8]
```diff
@@ -543,6 +543,39 @@ def __repr__(self) -> str:
...
+    ) -> None:
+        super().__init__()
+
+        assert isinstance(function, Expression)
+
+        assert (
+            isinstance(arguments, list)
+            and len(arguments) > 0
+            and all(isinstance(arg, ColumnReference) for arg in arguments)
+        )
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/column.py:560]: Adding these assertions here is helpful in the Python client but the server side does not do the same assertion. What happens if we drop the assertion on `ColumnReference` what would happen on the server? 

Is the analysis exception not better than the Python assertion>

File: python/pyspark/sql/connect/functions.py [hunk 9]
```diff
@@ -80,6 +84,80 @@ def _invoke_binary_math_function(name: str, col1: Any, col2: Any) -> Column:
...
+    # We should exclude functions that use
+    # variable args and keyword argnames
+    # as well as keyword only args
+    supported_parameter_types = {
+        inspect.Parameter.POSITIONAL_OR_KEYWORD,
+        inspect.Parameter.POSITIONAL_ONLY,
+    }
+
+    # Validate that
+    # function arity is between 1 and 3
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/functions.py:100]: one line?

File: python/pyspark/sql/connect/functions.py [hunk 10]
```diff
@@ -80,6 +84,80 @@ def _invoke_binary_math_function(name: str, col1: Any, col2: Any) -> Column:
     return _invoke_function(name, *_cols)
 
 
+def _get_lambda_parameters(f: Callable) -> ValuesView[inspect.Parameter]:
+    signature = inspect.signature(f)
+    parameters = signature.parameters.values()
+
+    # We should exclude functions that use
+    # variable args and keyword argnames
+    # as well as keyword only args
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/functions.py:93]:
```suggestion
    # We should exclude functions that use, variable args and keyword argument
    # names, as well as keyword only args.
```

File: python/pyspark/sql/connect/functions.py (hunk 9 above)
Comment (grundprinzip) [inline python/pyspark/sql/connect/functions.py:100]:
```suggestion
    # Validate that the function arity is between 1 and 3.
```

File: python/pyspark/sql/connect/functions.py [hunk 11]
```diff
@@ -80,6 +84,80 @@ def _invoke_binary_math_function(name: str, col1: Any, col2: Any) -> Column:
...
+    # Validate that
+    # function arity is between 1 and 3
+    if not (1 <= len(parameters) <= 3):
+        raise ValueError(
+            "f should take between 1 and 3 arguments, but provided function takes {}".format(
+                len(parameters)
+            )
+        )
+
+    # and all arguments can be used as positional
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/functions.py:108]:
```suggestion
    # Verify that all arguments can be used as positional arguments.
```

File: python/pyspark/sql/connect/functions.py [hunk 12]
```diff
@@ -80,6 +84,80 @@ def _invoke_binary_math_function(name: str, col1: Any, col2: Any) -> Column:
...
+    if not all(p.kind in supported_parameter_types for p in parameters):
+        raise ValueError("f should use only POSITIONAL or POSITIONAL OR KEYWORD arguments")
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/functions.py:110]: This error message is really hard to parse. I know what you want to say (and I know it's probably copy + paste from PySpark plain but still. 

Maybe better: "All arguments of f must be usable as POSITIONAL arguments."

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto (hunk 4 above)
Comment (grundprinzip) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto]: I think this is tricky because it's hard to provide clear guidance on what the right approach is. Generally, the downside of the unresolved function approach is that you're using a magic value that someone has to understand. This knowledge is now embedded in the client and cannot be inferred when looking at the protos.

Comment (zhengruifeng) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto]: will add LambdaFunction back in the protos

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 6 above)
Comment (zhengruifeng) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:569]: +1 will add it back

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 7 above)
Comment (zhengruifeng) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:586]: 1, the error message also mentioned `UnresolvedAttribute`

2, existing implementation in PySpark only use single part name

File: python/pyspark/sql/connect/functions.py (hunk 9 above)
Comment (zhengruifeng) [reply python/pyspark/sql/connect/functions.py]: done

File: python/pyspark/sql/connect/functions.py (hunk 12 above)
Comment (zhengruifeng) [reply python/pyspark/sql/connect/functions.py]: done

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto (hunk 4 above)
Comment (zhengruifeng) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:196]: > I think this is tricky because it's hard to provide clear guidance on what the right approach is. Generally, the downside of the unresolved function approach is that you're using a magic value that someone has to understand. This knowledge is now embedded in the client and cannot be inferred when looking at the protos.

Agreed, that is what I feel. We should avoid abusing unresolved function

File: python/pyspark/sql/connect/functions.py [hunk 13]
```diff
@@ -80,6 +84,78 @@ def _invoke_binary_math_function(name: str, col1: Any, col2: Any) -> Column:
...
+    to transformation described by f
+
+    :param f: A Python of one of the following forms:
+            - (Column) -> Column: ...
+            - (Column, Column) -> Column: ...
+            - (Column, Column, Column) -> Column: ...
+    """
+    parameters = _get_lambda_parameters(f)
+
+    arg_names = ["x", "y", "z"]
```
Comment (zhengruifeng) [inline python/pyspark/sql/connect/functions.py:125]: existing pyspark only uses single name part, so check it in the server side

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto (hunk 5 above)
Comment (zhengruifeng) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:203]: In the scala side, a LambdaFunction actually accepts `NamedExpression`s as argument;
In PySpark, an argument is always a `UnresolvedNamedLambdaVariable`;

Here we use `Expression` instead of `UnresolvedAttributes`, a benefit I can image is, if we need to support more types of Expression, we do not need to change the proto

File: python/pyspark/sql/connect/column.py (hunk 8 above)
Comment (zhengruifeng) [reply python/pyspark/sql/connect/column.py:556]: we also check `UnresolvedAttribute` in the server side

File: python/pyspark/sql/connect/functions.py (hunk 12 above)
Comment (zhengruifeng) [reply python/pyspark/sql/connect/functions.py]: done

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto (hunk 5 above)
Comment (cloud-fan) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:203]: Shall we simply use `repeated String ...`? If we think about the SQL API `exists(array_col, c -> ...)`, what we need to provide here is the argument name list, which is really just a list of string.

File: python/pyspark/sql/connect/functions.py [hunk 14]
```diff
@@ -80,6 +84,78 @@ def _invoke_binary_math_function(name: str, col1: Any, col2: Any) -> Column:
...
+    :param cols: a list of columns
+    :param funs: a list of((*Column) -> Column functions.
+
+    :return: a Column
+    """
+    assert len(funs) == 1
+    _cols = [_to_col(c) for c in cols]
+    _funs = [_create_lambda(f) for f in funs]
+
+    return _invoke_function(name, *_cols, *_funs)
```
Comment (cloud-fan) [inline python/pyspark/sql/connect/functions.py:156]: This looks like a hacky way to create lambda function...

File: python/pyspark/sql/connect/functions.py [hunk 15]
```diff
@@ -80,6 +84,78 @@ def _invoke_binary_math_function(name: str, col1: Any, col2: Any) -> Column:
...
+
+    :param f: A Python of one of the following forms:
+            - (Column) -> Column: ...
+            - (Column, Column) -> Column: ...
+            - (Column, Column, Column) -> Column: ...
+    """
+    parameters = _get_lambda_parameters(f)
+
+    arg_names = ["x", "y", "z"]
+    arg_cols = [column(arg) for arg in arg_names[: len(parameters)]]
```
Comment (cloud-fan) [inline python/pyspark/sql/connect/functions.py:126]: since we support up to 3 parameters, shall we add an assert here to avoid future mistakes?

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto (hunk 5 above)
Comment (zhengruifeng) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:203]: yes, that's simpler. At least `repeated String` is enough for existing pyspark's implementations. will update

File: python/pyspark/sql/connect/functions.py [hunk 16]
```diff
@@ -80,6 +84,78 @@ def _invoke_binary_math_function(name: str, col1: Any, col2: Any) -> Column:
...
+    # We should exclude functions that use, variable args and keyword argument
+    # names, as well as keyword only args.
+    supported_parameter_types = {
+        inspect.Parameter.POSITIONAL_OR_KEYWORD,
+        inspect.Parameter.POSITIONAL_ONLY,
+    }
+
+    # Validate that the function arity is between 1 and 3.
+    if not (1 <= len(parameters) <= 3):
+        raise ValueError(
```
Comment (zhengruifeng) [inline python/pyspark/sql/connect/functions.py:100]: @cloud-fan  Here is the validation on number of paramters

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 17]
```diff
@@ -534,6 +536,36 @@ class SparkConnectPlanner(session: SparkSession) {
     }
   }
 
+  /**
+   * Translates a LambdaFunction from proto to the Catalyst expression.
+   */
+  private def transformLamdaFunction(lambda: proto.Expression.LambdaFunction): Expression = {
+    if (lambda.getArgumentsCount == 0) {
```
Comment (grundprinzip) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:543]: Do we get an analysis exception if more than three arguments are submitted?

**Review by grundprinzip — COMMENTED** [review]
Looks good from my side.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 17 above)
Comment (amaliujia) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:543]: There is a validation in the client side already but I guess we need that in the server side as well?

**Review by amaliujia — APPROVED** [review]
LGTM

File: python/pyspark/sql/connect/functions.py [hunk 18]
```diff
@@ -80,6 +84,78 @@ def _invoke_binary_math_function(name: str, col1: Any, col2: Any) -> Column:
...
+            - (Column) -> Column: ...
+            - (Column, Column) -> Column: ...
+            - (Column, Column, Column) -> Column: ...
+    """
+    parameters = _get_lambda_parameters(f)
+
+    arg_names = ["x", "y", "z"][: len(parameters)]
+    arg_cols = [column(arg) for arg in arg_names]
+
+    result = f(*arg_cols)
```
Comment (cloud-fan) [inline python/pyspark/sql/connect/functions.py:128]: IIUC, we use `Column` with `UnresolvedAttribute` to invoke the function, then at the server side, we replace `UnresolvedAttribute` with `LambdaVariable`.

Shall we use `Column` with `LambdaVariable` to invoke the function in client side directly?

Comment (zhengruifeng) [reply python/pyspark/sql/connect/functions.py:128]: the first commit used `UnresolvedNamedLambdaVariable` 
```
    arg_cols: List[Column] = []
    for arg in arg_names[: len(parameters)]:
        # TODO: How to make sure lambda variable names are unique? RPC for increasing ID?
        _uuid = str(uuid.uuid4()).replace("-", "_")
        arg_cols.append(Column(UnresolvedNamedLambdaVariable([f"{arg}_{_uuid}"])))

    result = f(*arg_cols)

    if not isinstance(result, Column):
        raise ValueError(f"Callable {f} should return Column, got {type(result)}")

    return LambdaFunction(result._expr, [arg._expr for arg in arg_cols])
```

then switched to `UnresolvedAttribute` according to the comments https://github.com/apache/spark/pull/39068#discussion_r1049561414

Comment (cloud-fan) [reply python/pyspark/sql/connect/functions.py:128]: I don't think this can simplify the client-side implementation, but it does save one proto message. At least we should document this in the proto message `LambdaFunction`: the function body should use `UnresolvedAttribute` as arguments to build the query plan.

Comment (zhengruifeng) [reply python/pyspark/sql/connect/functions.py:128]: done

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by zhengruifeng** [comment]
thank you all for the reviews!
